wheel/python/clvm_rs/clvm_tree.py

```diff
@@ -0,0 +1,114 @@
+    deserialization, and keeps a reference to the serialized blob and to a
+    list of tuples of integers, each of which corresponds to a subtree.
+
+    It turns out every atom serialized to a blob contains a substring that
+    exactly matches that atom, so it ends up being not very wasteful to
+    simply use the blob for atom storage (especially if it's a `memoryview`,
+    from which you can take substrings without copying). Additionally, the
+    serialization for every object in the tree exactly corresponds to a
+    substring in the blob, so by clever caching we can very quickly generate
+    serializations for any subtree.
+
+    The deserializer iterates through the blob and caches a triple of
+    integers for each subtree: the first two integers represent the
+    `(start_offset, end_offset)` within the blob that corresponds to the
+    serialization of that object. You can check the contents of
+    `blob[start_offset]` to determine if the object is a pair (in which case
+    that byte is 0xff) or an atom (anything else). For a pair, the third
+    number corresponds to the index of the array that is the "rest" of the
+    pair (the "first" is always this object's index plus one, so we don't
+    need to save that); for an atom, the third number corresponds to an
+    offset of where the atom's binary data is relative to
+    `blob[start_offset]` (so the atom data is at `blob[triple[0] +
+    triple[2]:triple[1]]`)
+
+    Since each `CLVMTree` subtree keeps a reference to the original
+    serialized data and the list of tuples, no memory is released until all
+    objects in the tree are garbage-collected. This happens pretty naturally
+    in well-behaved python code.
+    """
+
+    # cached value of lazily-created child `CLVMStorage` objects
+    # created on call to `.pair`
+    _pair: Optional[Tuple["CLVMStorage", "CLVMStorage"]]
+
+    # the serialized form, which contains the template for all the atoms
+    blob: Union[memoryview, bytes]
+
+    # the memoized boundary list for all clvm objects in the tree
+    # this list is shared by *all* instances of sub-objects of the
+    # same tree
+    # [int, int, int] => start_offset, end_offset, atom_start_or_pair_rest
+    #   deserialization of this object is blob[start_offset, end_offset]
+    # We look at blob[start_offset] and if 0xff, this is a pair, otherwise atom
+    # if atom, atom is blob[atom_start_or_pair_rest:end_offset]
+    # if pair, children have `index` start_offset+1 and atom_start_or_pair_rest
+    int_tuples: List[Tuple[int, int, int]]
+
+    # which of the clvm objects in the tree this instance represents
+    index: int  # index into `int_tuples`
+
+    # the sha256 tree hashes that are optionally created on deserialization
+    tree_hashes: Optional[List[bytes]]
+
+    @classmethod
+    def from_bytes(cls, blob: bytes, calculate_tree_hash: bool = True) -> "CLVMTree":
+        int_tuples, tree_hashes = deserialize_as_tuples(
```

Comment:
you parse the whole tree structure eagerly here. In case the program is just run, this is a waste of time. It would seem more appropriate to do this lazily, don't you think?

You also compute the tree hashes eagerly by default, that also seems unintuitive to me. Surely that can't be a win on average, is it?

Reply:
My tests suggest the parsing will be fast enough to not be a big deal in most cases. Calculating the hash lazily is much slower because it has to be done in python (so we can check subtrees for cached hashes). Even if ported to rust, it would still have to context switch repeatedly into python to call `.atom` and `.pair` down the tree. At parse-time, we can stay in rust the entire time.

We could make `.parse` have a different default for `calculate_tree_hash` if this is implicitly called during consensus. Or have you completed your port to rust of deserializing these blobs?

For wallet use, deserialization will mostly be done on puzzle template `MOD` files, once per mod at start-up, and having the hashes is useful for doing fast zero-knowledge currying.

Opting out when performance is crucial is still possible, but I don't know if any cases of this still exist.

Reply:
> My tests suggest the parsing will be fast enough to not be a big deal in most cases.

From my experience of fixing performance issues in the full node and the wallet, deserializing objects is one of our main bottlenecks.

> Even if ported to rust, it would still have to context switch repeatedly into python to call .atom and .pair down the tree

I'm sure there are ways to avoid that, and stay in rust even when calculating it lazily.

> Or have you completed your port to rust of deserializing these blobs?

Everything has a rust counterpart, and it can be serialized and deserialized. The two main issues are:
1. the rust types are just the data, not all the functionality that's currently implemented as member functions. `Program` is also just a blob of bytes.
2. BLS types, G1Element, G2Element etc. are just byte blobs in rust, and it's quite expensive to parse them into the internal representation of relic, so that can't really be done on the fly. So to use the rust types in python requires a lot of changes to the class interfaces

> For wallet use, deserialization will mostly be done on puzzle template MOD files

Since those only happens once, on startup, I would think incoming Spends would be a much more common case for parsing puzzles and solutions.

> I just realized, the only implementation you see here is python. I have a rust version of the tree parsing that's not included in this patch.

Is there a good reason to have a python implementation at all? If there is a fast version, having a slow version seems risky, as someone might use it by accident.

Reply:
I wrote the python implementation first as a reference implementation. It's much easier to read and to understand. It's pretty hard to use it by accident.

If for some reason someone in future wants to run in pure python, this gives them a way to do so. I see rust as an accelerator, and whenever we can make it an optional feature, I think it makes sense to do so. I bet there is some marginal platform out there that has better support for python than rust.

Reply:
I can do further work down the line to separate out parsing and hashing, and in fact, I think it may make the code a lot simpler. I don't know if there will be a performance penalty, as we might have to load each atom into cache twice instead of once.

That said, do understand that merging this will have no impact on `chia-blockchain` until we actually do the next step of trying to replace `clvm` with `clvm_rs`. It might be more enlightening to run benchmarks on that PR and try to improve things then.